### PR TITLE
fix initialization in FastRandomContext: removes undefined behavior & uninitialized read

### PR DESCRIPTION
**Pull request by martinus**
When building #23152 with g++ 11.1 a few `-Wmaybe-uninitialized` warnings are shown. The warning in `FastRandomContext` seems legit, and this PR fixes that.

Note that there are more warnings in leveldb code which I did not investigate. There are warnings in fuzz/float.cpp and fuzz/string.cpp which are also left untouched (see https://github.com/bitcoin/bitcoin/pull/23169#discussion_r721084621)

### CODE REVIEW

File: src/test/fuzz/float.cpp [hunk 1]
```diff
@@ -19,7 +19,7 @@ FUZZ_TARGET(float)
 
     {
         const double d{[&] {
-            double tmp;
+            double tmp{0};
```
Comment (MarcoFalke) [inline src/test/fuzz/float.cpp:22]: The goal of leaving this uninitialized is to have memory sanitizers yell. By initializing you disable memory sanitizers.

File: src/test/fuzz/string.cpp [hunk 2]
```diff
@@ -224,22 +224,22 @@ FUZZ_TARGET(string)
...
-        int32_t i32;
-        int64_t i64;
-        uint32_t u32;
-        uint64_t u64;
-        uint8_t u8;
+        int32_t i32{0};
+        int64_t i64{0};
+        uint32_t u32{0};
+        uint64_t u64{0};
+        uint8_t u8{0};
```
Comment (MarcoFalke) [inline src/test/fuzz/string.cpp:231]: Same

File: src/test/fuzz/float.cpp (hunk 1 above)
Comment (MarcoFalke) [reply src/test/fuzz/float.cpp:22]: One workaround would be to make this optional, then `return *Assert(tmp);`. But I am not sure if this is worth it.

**Comment by practicalswift** [comment]
cr ACK 749e88eb86613932387afafac321df8822d9a951

File: src/test/fuzz/float.cpp (hunk 1 above)
Comment (martinus) [reply src/test/fuzz/float.cpp:22]: ah I didn't think about that. I'll remove this commit

**Comment by practicalswift** [comment]
Concept ACK on fixing the unitialized read (`bitbuf`) which can be tested with Valgrind using the following patch:

```patch
diff --git a/src/random.cpp b/src/random.cpp
index 174f4cef3..73e946783 100644
--- a/src/random.cpp
+++ b/src/random.cpp
@@ -693,13 +693,30 @@ FastRandomContext::FastRandomContext(bool fDeterministic) noexcept : requires_se
     rng.SetKey(seed.begin(), 32);
 }

+// Force Valgrind use-of-uninitialized memory (UUM) violation if `o` is uninitialized.
+//
+// As suggested by @guidovranken in https://github.com/bitcoin/bitcoin/issues/22064
+template<typename T>
+void ForceValgrindWarningIfUninitialized(const T& o) {
+    static_assert(std::is_trivially_copyable<T>::value);
+    FILE* f = fopen("/dev/null", "wb");
+    fwrite(&o, sizeof(o), 1, f);
+    fclose(f);
+}
+
 FastRandomContext& FastRandomContext::operator=(FastRandomContext&& from) noexcept
 {
+    ForceValgrindWarningIfUninitialized(from.requires_seed);
     requires_seed = from.requires_seed;
+    ForceValgrindWarningIfUninitialized(from.rng);
     rng = from.rng;
+    ForceValgrindWarningIfUninitialized(from.bytebuf);
     std::copy(std::begin(from.bytebuf), std::end(from.bytebuf), std::begin(bytebuf));
+    ForceValgrindWarningIfUninitialized(from.bytebuf_size);
     bytebuf_size = from.bytebuf_size;
+    ForceValgrindWarningIfUninitialized(from.bitbuf);
     bitbuf = from.bitbuf;
+    ForceValgrindWarningIfUninitialized(from.bitbuf_size);
     bitbuf_size = from.bitbuf_size;
     from.requires_seed = true;
     from.bytebuf_size = 0;
```

And running `valgrind src/test/test_bitcoin -t addrman_tests/addrman_simple`:

```
==22209== Memcheck, a memory error detector
…
Running 1 test case...
==22209== Syscall param write(buf) points to uninitialised byte(s)
==22209==    at 0x736A264: write (write.c:27)
==22209==    by 0x72E522C: _IO_file_write@@GLIBC_2.2.5 (fileops.c:1203)
==22209==    by 0x72E6FC0: new_do_write (fileops.c:457)
==22209==    by 0x72E6FC0: _IO_do_write@@GLIBC_2.2.5 (fileops.c:433)
==22209==    by 0x72E637F: _IO_file_close_it@@GLIBC_2.2.5 (fileops.c:136)
==22209==    by 0x72D83F6: fclose@@GLIBC_2.2.5 (iofclose.c:53)
==22209==    by 0xDA86B7: ForceValgrindWarningIfUninitialized<unsigned char [64]> (random.cpp:704)
==22209==    by 0xDA86B7: FastRandomContext::operator=(FastRandomContext&&) (random.cpp:713)
==22209==    by 0x9CFD67: Seed(FastRandomContext&) (setup_common.cpp:66)
==22209==    by 0x9D066B: SeedInsecureRand (setup_common.h:61)
==22209==    by 0x9D066B: BasicTestingSetup::BasicTestingSetup(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::vector<char const*, std::allocator<char const*> > const&) (setup_common.cpp:105)
==22209==    by 0x3AB6BE: addrman_simple (addrman_tests.cpp:166)
…
```

**Comment by laanwj** [comment]
As it has historically been a source of some very serious vulnerabilities I'm somewhat concerned around fixing uninitialized value warnings in random functionality. Please review this carefully.
Concept ACK of course.

**Comment by MarcoFalke** [comment]
Yeah, would have been good if `FastRandomContext` used this from the beginning. For reference, chacha20 FastRandom was introduced in #9792

File: src/random.h [hunk 3]
```diff
@@ -119,14 +119,14 @@ void RandAddEvent(const uint32_t event_info) noexcept;
...
 {
 private:
-    bool requires_seed;
-    ChaCha20 rng;
+    bool requires_seed{true};
+    ChaCha20 rng{};
 
-    unsigned char bytebuf[64];
-    int bytebuf_size;
+    unsigned char bytebuf[64] = {};
```
Comment (MarcoFalke) [inline src/random.h:125]: shouldn't this be left uninitialized? With this patch memory sanitizer will fail to detect logic errors.

**Comment by laanwj** [comment]
Code review ACK f46225ea5cadcb94cbaf7b77346bfef622d35eed

**Comment by MarcoFalke** [comment]
I still don't understand this.

Reading https://github.com/bitcoin/bitcoin/pull/23169#issuecomment-933298687 tells me that the issue is (also) bytebuf, but you are fixing bitbuf and leave bytebuf an uninitialized  read?

**Comment by martinus** [comment]
My change was only about the undefined behavior issue, because assigning from an uninitialized variable that's not a `char` is undefined behavior. As far as I understand it that's not with bytebuf that's not undefined behavior but just reading uninitialized memory: https://eel.is/c++draft/basic.indet#example-1

I've rebased and updated the PR though to get rid of reading uninitialized memory as well in 10aeb6e

**Comment by sipa** [comment]
If the buffer sizes are initialized, I'm pretty sure the uninitialized area of the byte buffer itself will never be read from (because it will be filled before any read happens). So I'm in favor of not initializing the byte buffer, giving analysis tools a chance to detect bugs with it.

**Comment by MarcoFalke** [comment]
Jup, I agree, see https://github.com/bitcoin/bitcoin/pull/23169#discussion_r721261844.


Now that you removed the initialization, the pull request title etc should be updated? (The current patch does not change initialization)

**Comment by martinus** [comment]
Right, I've updated the title

**Comment by achow101** [comment]
cc @jonatack @LarryRuane

**Review by maflcko — COMMENTED** [review]
Can still reproduce with gcc 12.2

Left a question.

**Review by maflcko — APPROVED** [review]
Happy to re-ACK if you decide to change something

ACK 10aeb6e6484048b061620d484dbb10dee0ead386 🌻

<details><summary>Show signature</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

ACK 10aeb6e6484048b061620d484dbb10dee0ead386 🌻
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUi1fAv/Y3d+GlbG+mm4/OqUCynfgYTZQRKI6YGchUigjXm4pKVT+zMjUZ46r+CK
4ncyI81dSp4XEroQ4Ppstljk+6tYLq/1+E6usfMQnO109oFfTqkYVLBeMdqZBUSH
w7SMoQkEmHEYxrNbx08qAy3uyymnwqkJgS7JUfVsrmdhw7CvyPBICDzGkYrtPCx2
z13VVmQPMHWjDH5700pQp9zbAXBBba+WQBQ0UfsahdPBGTABDQx86x3PtKMyW4kI
LI7LTA1s5F6VCR2W53esqdJGGsn4D03l0LBkOKU5lwXncepfiMmncisOl4y2e2ZQ
J2/TsCyk1PjqitOnHJd8SlDxOP26EqfvtjH6IHK4SE3xTvnZZegj2UDRTXH3zmis
HS4VZBRF+nYWC94Nk8zYi+PwYoreyjTZy0K9YEJcb5T1QswZeyT/vEgm3fEEe85E
a0W0k5C0T27mzcy+W3G0W0h9zot+jjsPwwLQ+HC84QqwS3FtzxlOrtCI7rs1DzgE
mkX4AlCj
=Qwgq
-----END PGP SIGNATURE-----
```


</details>

**Comment by fanquake** [comment]
>  is not relevant any more / was fixed (as far as I can say)

There are still some `-Wmaybe-uninitialized` warnings produced with GCC 13.0.1 and `-flto`, some in random.cpp. Opened #27343 to document them + others.

**Pull request closed without merging**